utilities/src/main/java/ru/study21/jcsv/xxl/io/RandomAccessCachedNioReader.java

```diff
@@ -0,0 +1,88 @@
+package ru.study21.jcsv.xxl.io;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.Random;
+
+// assumes underlying file is not updated
+public class RandomAccessCachedNioReader {
+
+    private final ByteBuffer[] caches;
+    private final long[] positions;
+    private final FileChannel channel;
+    private final CachePolicy cachePolicy;
+
+    public interface CachePolicy {
+        int readCache(long pos) throws IOException; // return index of new cache
+
+        default void didRead(int cacheIndex) { /* do nothing */ } // trigger to calc stats (for LRU / ...)
+
+        enum Type {RR} // TODO: LRU requires heap with decrease-key, use already implemented PairingHeap
+    }
+
+    private class RRPolicy implements CachePolicy {
+        private final Random random = new Random(5);
+
+        @Override
+        public int readCache(long pos) throws IOException {
+            int index = random.nextInt(caches.length);
+            caches[index].clear();
+            channel.read(caches[index], pos); // might not fill to capacity!
+            positions[index] = pos;
+            return index;
+        }
+    }
+
+    public RandomAccessCachedNioReader(
+            FileChannel channel,
+            long approxMemoryLimit,
+            int cacheCount, // defines cache size,
+            CachePolicy.Type cacheType
+    ) {
+        this.channel = channel;
+
+        if (approxMemoryLimit / cacheCount > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("too large cache size");
+        }
+        int cacheSize = (int) (approxMemoryLimit / cacheCount);
+        caches = new ByteBuffer[cacheCount];
+        positions = new long[cacheCount];
+        for (int i = 0; i < cacheCount; i++) {
+            caches[i] = ByteBuffer.allocate(cacheSize);
+            positions[i] = Long.MAX_VALUE; // ensure no cache works until written at least once
+        }
+
+        cachePolicy = switch (cacheType) {
+            case RR -> new RRPolicy();
+            // case LRU -> new LRUPolicy();
+            default -> throw new IllegalArgumentException("unknown cache type");
+        };
+    }
+
+    public void read(byte[] arr, long pos) throws IOException {
+        int cacheIndex = fittingCache(pos, arr.length);
+        if (cacheIndex == -1) {
+            cacheIndex = cachePolicy.readCache(pos);
+        }
```

Comment:
What about such sequence of pos: x, x - 100, x - 200, x - 300 (some slice of small records that goes to reversed order).

It will lead to wasting cache for a lot of overlapping areas.

It would be better to have fixed size pages in cache and load two pages if some record passes pages border.

It could be one of reasons for poor performance at level of java code (together with lack of LRU).